store.go

```diff
@@ -132,3 +132,37 @@ func (this *Storage) getPage(page string) map[string]*Socket {
 	this.pageMu.RLock()
 	return this.memory.getPage(page)
 }
+
+func (this *Storage) SetGroups(sock *Socket) error {
+	this.pageMu.Lock()
+	this.memory.SetGroups(sock)
+	this.pageMu.Unlock()
+
+	if this.StorageType == "redis" {
+		if err := this.redis.SetGroups(sock); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (this *Storage) UnsetGroups(sock *Socket) error {
+	this.pageMu.Lock()
+	this.memory.UnsetGroups(sock)
+	this.pageMu.Unlock()
+
+	if this.StorageType == "redis" {
+		if err := this.redis.UnsetPage(sock); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (this *Storage) getGroups(groups []string) []map[string]*Socket {
+	defer this.pageMu.RUnlock()
```

Comment:
Same, should use a different lock. Also could you move the defer statement below the `Lock()` call?